sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/RequiresStableInputIT.java

```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Integration test for the support of {@link
+ * org.apache.beam.sdk.transforms.DoFn.RequiresStableInput} annotation.
+ */
+@RunWith(JUnit4.class)
+public class RequiresStableInputIT {
+
+  private static final String VALUE = "value";
+  // SHA-1 hash of string "value"
+  private static final String VALUE_CHECKSUM = "f32b67c7e26342af42efabc674d441dca0a281c5";
+
+  private static class PairWithRandomKeyFn extends SimpleFunction<String, KV<String, String>> {
+    @Override
+    public KV<String, String> apply(String value) {
+      String key = UUID.randomUUID().toString();
+      return KV.of(key, value);
+    }
+  }
+
+  private static class MakeSideEffectAndThenFailFn extends DoFn<KV<String, String>, String> {
+    private final String outputPrefix;
+
+    private MakeSideEffectAndThenFailFn(String outputPrefix) {
+      this.outputPrefix = outputPrefix;
+    }
+
+    @RequiresStableInput
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      MatchResult matchResult = FileSystems.match(outputPrefix + "*");
+      boolean firstTime = (matchResult.metadata().size() == 0);
+
+      KV<String, String> kv = c.element();
+      writeTextToFileSideEffect(kv.getValue(), outputPrefix + kv.getKey());
+      if (firstTime) {
+        throw new Exception("Deliberate failure: should happen only once.");
+      }
+    }
+
+    private static void writeTextToFileSideEffect(String text, String filename) throws IOException {
+      ResourceId rid = FileSystems.matchNewResource(filename, false);
+      WritableByteChannel chan = FileSystems.create(rid, "text/plain");
+      chan.write(ByteBuffer.wrap(text.getBytes(Charset.defaultCharset())));
+      chan.close();
+    }
+  }
+
+  private static void runRequiresStableInputPipeline(RequiresStableInputTestOptions options) {
+    Pipeline p = Pipeline.create(options);
+
+    PCollection<String> singleton = p.apply("CreatePCollectionOfOneValue", Create.of(VALUE));
+    singleton
+        .apply("Single-PairWithRandomKey", MapElements.via(new PairWithRandomKeyFn()))
+        .apply(
+            "Single-MakeSideEffectAndThenFail",
+            ParDo.of(new MakeSideEffectAndThenFailFn(options.getParDoSingleOutputPrefix())));
+    singleton
+        .apply("Multi-PairWithRandomKey", MapElements.via(new PairWithRandomKeyFn()))
+        .apply(
+            "Multi-MakeSideEffectAndThenFail",
+            ParDo.of(new MakeSideEffectAndThenFailFn(options.getParDoMultiOutputPrefix()))
+                .withOutputTags(new TupleTag<>(), TupleTagList.empty()));
+
+    p.run().waitUntilFinish();
+  }
+
+  @BeforeClass
+  public static void setup() {
+    PipelineOptionsFactory.register(TestPipelineOptions.class);
+  }
+
+  /**
+   * Option for RequiresStableInputIT.
+   *
+   * <p>Define the output prefixes for tests on {@link ParDo.SingleOutput} and {@link
+   * ParDo.MultiOutput}.
+   */
+  public interface RequiresStableInputTestOptions extends TestPipelineOptions {
```

Comment:
Instead of needing an options class, why not just pass this into the DoFn as a property that gets serialized at pipeline construction time?